models/repo_transfer.go

```diff
@@ -0,0 +1,237 @@
+	Rejected RepoStatus = iota
+	// Pending is the default repo transfer state. All initiated transfers
+	// automatically get this status.
+	Pending
+	// Accepted is a repo transfer state for repository transfers that have
+	// been acknowledged by the recipient
+	Accepted
+)
+
+// RepoTransfer is used to manage repository transfers
+type RepoTransfer struct {
+	ID          int64 `xorm:"pk autoincr"`
+	RecipientID int64
+	Recipient   *User `xorm:"-"`
+	RepoID      int64
+	CreatedUnix util.TimeStamp `xorm:"INDEX NOT NULL created"`
+	UpdatedUnix util.TimeStamp `xorm:"INDEX NOT NULL updated"`
+	Status      RepoStatus
+}
+
+// LoadRecipient fetches the transfer recipient from the database
+func (r *RepoTransfer) LoadRecipient() error {
+	if r.Recipient != nil {
+		return nil
+	}
+
+	u, err := GetUserByID(r.RecipientID)
+	if err != nil {
+		return err
+	}
+
+	r.Recipient = u
+	return nil
+}
+
+// GetPendingRepositoryTransfer fetches the most recent and ongoing transfer
+// process for the repository
+func GetPendingRepositoryTransfer(repo *Repository) (*RepoTransfer, error) {
+	var transfer = new(RepoTransfer)
+
+	_, err := x.Where("status = ? AND repo_id = ? ", Pending, repo.ID).
+		Get(transfer)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if transfer.ID == 0 {
+		return nil, ErrNoPendingRepoTransfer{RepoID: repo.ID}
+	}
+
+	return transfer, nil
+}
+
+// CancelRepositoryTransfer makes sure to set the transfer process as
+// "rejected". Thus ending the transfer process
+func CancelRepositoryTransfer(repoTransfer *RepoTransfer) error {
+	repoTransfer.Status = Rejected
+	repoTransfer.UpdatedUnix = util.TimeStampNow()
+	_, err := x.ID(repoTransfer.ID).Cols("updated_unix", "status").
+		Update(repoTransfer)
+	return err
+}
+
+// StartRepositoryTransfer marks the repository transfer as "pending". It
+// doesn't actually transfer the repository until the user acks the transfer.
+func StartRepositoryTransfer(newOwnerName string, repo *Repository) error {
+	// Make sure the repo isn't being transferred to someone currently
+	// Only one transfer process can be initiated at a time.
+	// It has to be cancelled for a new transfer to occur
+
+	n, err := x.Count(&RepoTransfer{
+		RepoID: repo.ID,
+		Status: Pending,
+	})
+	if err != nil {
+		return err
+	}
+
+	if n > 0 {
+		return ErrRepoTransferInProgress{newOwnerName, repo.Name}
+	}
+
+	newOwner, err := GetUserByName(newOwnerName)
+	if err != nil {
+		return fmt.Errorf("get new owner '%s': %v", newOwnerName, err)
+	}
+
+	// Check if new owner has repository with same name.
+	has, err := IsRepositoryExist(newOwner, repo.Name)
+	if err != nil {
+		return fmt.Errorf("IsRepositoryExist: %v", err)
+	} else if has {
+		return ErrRepoAlreadyExist{newOwnerName, repo.Name}
+	}
+
+	transfer := &RepoTransfer{
+		RepoID:      repo.ID,
+		RecipientID: newOwner.ID,
+		Status:      Pending,
+		CreatedUnix: util.TimeStampNow(),
+		UpdatedUnix: util.TimeStampNow(),
+	}
+
+	_, err = x.Insert(transfer)
```

Comment:
And can user still push codes or something on this repository?

Reply:
I will still have to look into his Github does this but my instinct says pushing should still be allowed

Reply:
Yes, code can still go into the repository.. I will add a redirect though